TESTS/network/emac/emac_TestNetworkStack.cpp

```diff
@@ -108,8 +108,10 @@ nsapi_error_t EmacTestNetworkStack::add_ethernet_interface(EMAC &emac, bool defa
 {
     // Test network stack supports only one interface
     TEST_ASSERT_MESSAGE(!m_interface, "Only one interface supported!");
-
-    m_interface = &EmacTestNetworkStack::Interface::get_instance();
+#ifdef TARGET_UBLOX_EVK_ODIN_W2
+	emac_if_init();
```

Comment:
There should be no need to modify the test for this. If you need to internally power up earlier than the "power up" you are free to do that internally.

Only thing you might need to do is make sure you issue "Link up" from inside the actual `EMAC::power_up` if you're already up.

Reply:
BTW, be aware of the fix #6296 - that avoids a problem with issuing link up from inside `power-up`.

Reply:
Could please elaborate it more? I havn't find anything in 
https://github.com/ARMmbed/mbed-os/pull/6296

In case of lwip emac_if_init called in a add_ethernet_interface but in test cases its quite different. odin interface need this emac initialization before the bring_up call for wifi connection messages.

this is a call stack with lwip

	LWIP::Interface::emac_if_init(netif = 0x20018734)	C/C++
 	netif_add(netif = 0x20018734, ipaddr = 0x0, netmask = 0x0, gw = 0x0, state = 0x200186d8, init = 0x80039fd <LWIP::Interface::emac_if_init(netif*)>, input = 0x8007975 <tcpip_input>)	C/C++
 	LWIP::add_ethernet_interface(this = 0x20001848 <_ZZN4LWIP12get_instanceEvE4lwip>, emac = @0x20002018: {_vptr.EMAC = 0x8098460 <_ZTV9WIFI_EMAC+8>}, default_if = true, interface_out = 0x20011e78)	C/C++
 	OdinWiFiInterface::handle_wlan_status_started(this = 0x20011e68, start = 0x200126c0)	C/C++
 	OdinWiFiInterface::handle_in_msg(this = 0x20011e68)	C/C++
 	wlan_callb_s::odin_thread_fcn(wifi = 0x20011e68)	C/C++
 	mbed::Callback<void ()>::function_context<void (*)(OdinWiFiInterface*), OdinWiFiInterface>::operator()() const(this = 0x200120e8)	C/C++
 	mbed::Callback<void ()>::function_call<mbed::Callback<void ()>::function_context<void (*)(OdinWiFiInterface*), OdinWiFiInterface> >(void const*)(p = 0x200120e8)	C/C++
 	mbed::Callback<void ()>::call() const(this = 0x200120e8)	C/C++
 	mbed::Callback<void ()>::operator()() const(this = 0x200120e8)	C/C++
 	rtos::Thread::_thunk(thread_ptr = 0x200120e4)	C/C++

And it will be only satisfied by this change

	WIFI_EMAC::set_memory_manager(this = 0x20002874 <_ZZN9WIFI_EMAC12get_instanceEvE4emac>, mem_mngr = @0x20001114: {_vptr.EMACMemoryManager = 0x80887d4 <_ZTV21EmacTestMemoryManager+8>})	C/C++
 	emac_if_init()	C/C++
 	EmacTestNetworkStack::add_ethernet_interface(this = 0x20001160 <_ZZN20EmacTestNetworkStack12get_instanceEvE10test_stack>, emac = @0x20002874: {_vptr.EMAC = 0x808c8f0 <_ZTV9WIFI_EMAC+8>}, default_if = true, interface_out = 0x200134f0)	C/C++
 	OdinWiFiInterface::handle_wlan_status_started(this = 0x200134e0, start = 0x20013858)	C/C++
 	OdinWiFiInterface::handle_in_msg(this = 0x200134e0)	C/C++
 	wlan_callb_s::odin_thread_fcn(wifi = 0x200134e0)	C/C++
 	mbed::Callback<void ()>::function_context<void (*)(OdinWiFiInterface*), OdinWiFiInterface>::operator()() const(this = 0x20013760)	C/C++
 	mbed::Callback<void ()>::function_call<mbed::Callback<void ()>::function_context<void (*)(OdinWiFiInterface*), OdinWiFiInterface> >(void const*)(p = 0x20013760)	C/C++
 	mbed::Callback<void ()>::call() const(this = 0x20013760)	C/C++
 	mbed::Callback<void ()>::operator()() const(this = 0x20013760)	C/C++
 	rtos::Thread::_thunk(thread_ptr = 0x2001375c)	C/C++

Reply:
You can internally power yourself up whenever you need to.  You could have your `NetworkInterface::connect` do whatever initialisation you need yourself. See for example #6904, which performs its first `wifi_connect` powerup before calling either `add_network_interface` or `bringup`.

Your actual `EMAC::powerup` can be basically null, from your hardware's point of view, if you are already bringing the interface up for the "control" paths. The only issue is that if you're already up, `powerup` marks the point at which you should be (re)-sending the "link up" call back, as the stack will not be expecting it before then.

And #6926 (sorry, not 6296 - my typo there) corrects the logic in lwIP so it doesn't slip up if the "link up" is called from inside powerup.

Reply:
In LWIP and tests,  emac_if_init  sets the emac memory manager "emac->set_memory_manager(*memory_manager);" and ODIN interface allocate messages by this memory manager for connection setup. 

>>  You can internally power yourself up whenever you need to.

The actual problem is the allocation of those messages by using *memory_manager and its only possible by calling emac_if_init(); before the bring_up call which is .

	nsapi_error_t EmacTestNetworkStack::Interface::bringup(bool dhcp, const char *ip, const char *netmask, const char *gw, const nsapi_ip_stack_t stack, bool blocking)
	{
		if (!emac_if_init()) {
			TEST_ASSERT_MESSAGE(0, "emac initialization failed!");
		}

		return NSAPI_ERROR_OK;
	}

@andreaslarssonublox

Reply:
Got it.

How about if we guarantee that the memory manager is registered in the "add_ethernet_interface"? Would that suffice?

Reply:
Not sure what do you mean by that. If it will set this

    void WIFI_EMAC::set_memory_manager(EMACMemoryManager &mem_mngr)
    {
        memory_manager = &mem_mngr;
    }

it is sufficent.

Reply:
I made a pull request to add the set_memory_manager() call to test stack Ethernet interface add function:

https://github.com/ARMmbed/mbed-os/pull/6937

Will that fix the problem?

Reply:
Yes, it worked.

Reply:
Great. If you can give us an updated PR without the test modification (so it that relies on #6937 being present), and addresses the other comments, I can kick off combined CI to get the first pass integrated.